bin/ch-run.c

```diff
@@ -103,191 +59,81 @@ const struct argp_option options[] = {
    { 0 }
 };
 
-struct bind {
-   char * src;
-   char * dst;
-};
-
 struct args {
-   struct bind binds[USER_BINDS_MAX+1];
-   gid_t container_gid;
-   uid_t container_uid;
-   char * newroot;
+   struct container c;
    char * initial_dir;
-   bool join;
-   int join_ct;
-   char * join_tag;
-   bool private_home;
-   bool private_tmp;
-   bool writable;
-   int user_cmd_start;  // index into argv where NEWROOT is
 };
 
-void enter_udss(char * newroot, bool writeable, struct bind * binds,
-                bool private_tmp, bool private_home);
+/** Function prototypes **/
+
 bool get_first_env(char ** array, char ** name, char ** value);
-void join_begin();
-int join_ct();
-void join_end();
-void join_namespace(pid_t pid, char * ns);
-void join_namespaces();
-char * join_tag();
-void log_ids(const char * func, int line);
-void run_user_command(int argc, char * argv[], int user_cmd_start);
-static error_t parse_opt(int key, char * arg, struct argp_state * state);
+int join_ct(int cli_ct);
+char * join_tag(char * cli_tag);
 int parse_int(char * s, bool extra_ok, char * error_tag);
+static error_t parse_opt(int key, char * arg, struct argp_state * state);
 void privs_verify_invoking();
```

Comment:
Personally, I would move ch-run's prototypes to charliecloud.h.

Reply:
They're in `ch-run.c` because that's the only place they're used. Prototypes in `charliecloud.h` should be applicable to anyone who uses `charliecloud.c`, which may eventually be more than `ch-run.c`.